test/Directory.Build.targets

```diff
@@ -1,5 +1,13 @@
 <Project>
 
   <Import Project="..\Directory.Build.targets" />
-
+  
+  <PropertyGroup Condition="'$(Coverage)' == 'true' and '$(Configuration)'=='Release' and '$(OS)' == 'Windows_NT' and '$(TargetArchitecture)' == 'x64'">
```

Comment:
A few questions about this -

1. Why is this being set in a `.targets` file? It should go into the `.props` file instead.
2. Why only on Windows and Release builds?  (I assume the x64 is a limitation of the coverage tool. If it isn't - then why the condition on it?)

Reply:
It belongs into the targets file as it relies on AssemblyName which isn't available before the targets file. I saw that this was now changed to MSBuildProjectName which isn't correct.

Reply:
Why isn’t it correct?

Reply:
You can override an assembly's name with the `<AssemblyName>` property. If you build a project the produced assembly will use that name and not the project name. Hence the coverage report filename should also rely on `AssemblyName`.

Reply:
In ML.NET, the `$(MSBuildProjectName)` will always equal `$(AssemblyName)`. If someone wants to stray from that pattern, they can set the `CoverletOutput` property as well. But it would be highly discouraged to not keep the project name match the assembly name.

Reply:
Ok but what's the *reason* why you wouldn't use AssemblyName here?

Reply:
The reason is so these properties can be set in a `.props` file. Setting properties in `.targets` files is discouraged (for me anyway) because then the `.csproj` can't override them without dropping out of the `<Project Sdk="Microsoft.NET.Sdk">` syntax.

Reply:
Another question to be asking here is: Why doesn't Coverlet default this for us? Why are we (and every other consumer) responsible for setting this?

Reply:
I know that you have a lot of experience in this area but I wouldn't encourage others to follow this principle. In this case it works as there's a msbuild property that could be used before the project is loaded but usually that's not the case. Also by conditioning these properties in a targets file with `'$(Property)' == ''` allows you to overwrite them in the project without dropping the sdk attribute.

cc @ericstj for opinion.

Reply:
> Another question to be asking here is: Why doesn't Coverlet default this for us? Why are we (and every other consumer) responsible for setting this?

I believe @codemzs mentioned that otherwise the file would just be dropped in the project directory.

Reply:
> Another question to be asking here is: Why doesn't Coverlet default this for us? Why are we (and every other consumer) responsible for setting this?

It does set a default: `$(MSBuildProjectDirectory)coverage\`

Reply:
Who wants that default? Outputs shouldn't be dropped in the source directory.... They should be going to the output directory.

Looking at the code, it also doesn't output a file based on the `$(AssemblyName)` either - https://github.com/tonerdo/coverlet/blob/28fb5af1947c7240a7c36ebc4072daa0c80d26d5/src/coverlet.msbuild.tasks/CoverageResultTask.cs#L95.  What would be wrong with outputting the file to `bin\AnyCPU.Debug\Microsoft.ML.Core.Tests\netcoreapp2.1\coverage.opencover.xml`?
In order to do that, you can just set `<CoverletOutput>$(OutputPath)</>`

Reply:
> Who wants that default? Outputs shouldn't be dropped in the source directory.... They should be going to the output directory.

Please open an issue for that in https://github.com/tonerdo/coverlet/.

> What would be wrong with outputting the file to 

+1.

Reply:
@ViktorHofer - we do the same thing in corefx (assuming `MSBuildProjectName == AssemblyName`). It would be very confusing if people didn't follow this convention:

https://github.com/dotnet/corefx/blob/77096755443730962810662501f82550fdb4656d/Directory.Build.props#L478
https://github.com/dotnet/corefx/blob/77096755443730962810662501f82550fdb4656d/Directory.Build.props#L356
https://github.com/dotnet/corefx/blob/8f2afd786b04ba95b13d5a50b6f1e566b16cf5ac/eng/ReferenceAssemblies.props#L11-L12
https://github.com/dotnet/corefx/blob/8735f858c09443b52a25a48c6f5b9aaf01e78043/src/shims/manual/Directory.Build.props#L29

Reply:
> Please open an issue for that in https://github.com/tonerdo/coverlet/.

There already is an issue https://github.com/tonerdo/coverlet/issues/152, but it was closed without a reason/justification. I've asked why it was closed.